lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -183,84 +87,127 @@ public void Dispose()
+                                return fun(...args);
+                                }, ms);
+                            }",
+                            new object[]
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _pollingInterval,
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
+                }
+                else if (_polling == WaitForFunctionPollingOption.Raf)
+                {
+                    _poller = await _isolatedWorld.EvaluateFunctionHandleAsync(
+                            @"
+                            ({RAFPoller, createFunction}, fn, ...args) => {
+                                const fun = createFunction(fn);
+                                return new RAFPoller(() => {
+                                    return fun(...args);
+                                });
+                            }",
+                            new object[]
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
+                }
+                else
                 {
-                    await success.DisposeAsync().ConfigureAwait(false);
+                    _poller = await _isolatedWorld.EvaluateFunctionHandleAsync(
+                            @"
+                            ({MutationPoller, createFunction}, root, fn, ...args) => {
+                                const fun = createFunction(fn);
+                                return new MutationPoller(() => {
+                                return fun(...args);
+                                }, root || document);
+                            }",
+                            new object[]
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _root,
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
                 }
 
-                return;
-            }
+                await _poller.EvaluateFunctionAsync("poller => poller.start()").ConfigureAwait(false);
 
-            if (exception == null &&
-                await _isolatedWorld.EvaluateFunctionAsync<bool>("s => !s", success)
-                    .ContinueWith(
-                        task => task.IsFaulted || task.Result,
-                        TaskScheduler.Default)
-                    .ConfigureAwait(false))
+                var success = await _poller.EvaluateFunctionHandleAsync("poller => poller.result()").ConfigureAwait(false);
+                _result.TrySetResult(success);
+                await TerminateAsync().ConfigureAwait(false);
```

Comment:
If we dispose `_poller` here anyways, it seems we could keep `_poller` a local variable?
Less mutable state just makes thread-safety easier.

It could look something like this

```cs
internal async Task Rerun()
{
    IJSHandle poller = null;

    try
    {
        var context = await _isolatedWorld.GetExecutionContextAsync().ConfigureAwait(false);
        await System.Threading.Tasks.Task.WhenAll(_bindings.Select(binding => _isolatedWorld.AddBindingToContextAsync(context, binding.Name))).ConfigureAwait(false);

        if (_pollingInterval.HasValue)
        {
            poller = [...]
        }
        else if (_polling == WaitForFunctionPollingOption.Raf)
        {
            poller = [...]
        }
        else
        {
            poller = [...]
        }

        await poller.EvaluateFunctionAsync("poller => poller.start()").ConfigureAwait(false);

        var success = await poller.EvaluateFunctionHandleAsync("poller => poller.result()").ConfigureAwait(false);
        _result.TrySetResult(success);
        await Terminate(poller).ConfigureAwait(false);
        Terminate();
    }
    catch (Exception ex)
    {
        if (poller is not null)
        {
            await Terminate(poller).ConfigureAwait(false);
        }

        var exception = GetBadException(ex);
        if (exception != null)
        {
            Terminate(exception);
        }
    }
}

internal void Terminate(Exception exception = null)
{
    // The timeout timer might call this method on cleanup
    if (_terminated)
    {
        return;
    }

    _terminated = true;
    _isolatedWorld.TaskManager.Delete(this);
    Cleanup(); // This matches the clearTimeout upstream

    if (exception != null)
    {
        _result.TrySetException(exception);
    }
}

private async Task Terminate(IJSHandle poller)
{
    try
    {
        await using (poller.ConfigureAwait(false))
        {
            await poller.EvaluateFunctionAsync(@"async poller => {
                await poller.stop();
            }").ConfigureAwait(false);
        }
    }
    catch (Exception)
    {
        // swallow error.
    }
}
```

Reply:
Wait... I just saw the case when `_poller` has not completed yet, but `TerminateAsync` should terminate it anyways.

I wanted to guard against repeated calls to `Rerun` overwriting `_poller` as that could mean that only the _last_ poller is disposed. 

It's tricky 😃